ipatests/test_ipaserver/test_adtrust_mockup.py

```diff
@@ -0,0 +1,59 @@
+# Copyright (C) 2018  FreeIPA Project Contributors - see LICENSE file
+
+from __future__ import print_function
+import ipaserver.install.adtrust as adtr
+from ipaserver.install.adtrust import set_and_check_netbios_name
+from collections import namedtuple
+from unittest import TestCase
+from unittest.mock import MagicMock
+import sys
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
+
+
+class ApiMockup:
+    Backend = namedtuple('Backend', 'ldap2')
+    Calls = namedtuple('Callbacks', 'retrieve_netbios_name')
+    env = namedtuple('Environment', 'domain')
+
+
+class TestNetbiosName(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        api = ApiMockup()
+        ldap2 = namedtuple('LDAP', 'isconnected')
+        ldap2.isconnected = MagicMock(return_value=True)
+        api.Backend.ldap2 = ldap2
+        api.Calls.retrieve_netbios_name = adtr.retrieve_netbios_name
+        adtr.retrieve_netbios_name = MagicMock(return_value=None)
+        cls.api = api
+
+    @classmethod
+    def tearDownClass(cls):
+        adtr.retrieve_netbios_name = cls.api.Calls.retrieve_netbios_name
+
+    def test_NetbiosName(self):
+        """
+        Test set_and_check_netbios_name() using permutation of two inputs:
+        - predefined and not defined NetBIOS name
+        - unattended and interactive run
+        As result, the function has to return expected NetBIOS name in
+        all cases. For interactive run we override input to force what
+        we expect.
+        """
+        self.api.env.domain = 'example.com'
+        expected_nname = 'EXAMPLE'
+        # NetBIOS name, unattended, should set the name?
+        tests = ((expected_nname, True, False),
+                 (None, True, True),
+                 (None, False, True),
+                 (expected_nname, False, False))
+        sys.stdin = StringIO(expected_nname + '\r')
```

Comment:
Since you already started to use the mock module, how about ``mock.patch`` here? You can use it either as a context manager or a function decorator.

```
>>> import sys
>>> import unittest.mock
>>> from io import StringIO
>>> with unittest.mock.patch('sys.stdin', new_callable=StringIO):
...     print(sys.stdin)
... 
<_io.StringIO object at 0x7f438ac23ca8>
>>> print(sys.stdin)
<_io.TextIOWrapper name='<stdin>' mode='r' encoding='UTF-8'>
>>> @unittest.mock.patch('sys.stdin', new_callable=StringIO)
... def example(m_stdin):
...     print(sys.stdin)
... 
>>> example()
<_io.StringIO object at 0x7f438a9e5708>
```